File: Data algorithms and structures/02 list/presidents.py
import pandas as pd 
import numpy as np

class President:
    def __init__(self, data):
        self.name = data[0]
        self.years = data[1]
        self.party = data[2]
        self.end = data[3]
        self.start = data[4]

    def show(self):
        print('Name: ', self.name)
        print('RuleYears: ', self.years)
        print('Party: ', self.party)
        print('Start: ', self.start)
        print('End: ', self.end)

    def get_data(self):
        return self.name, self.years, self.party, self.end, self.start

class Object:
    def __init__(self, data):
        self.data = President(data)
        self.next = None
        self.prev = None
        self.index = 0
    
    def show(self):
        print('index: ', self.index)
        self.data.show()
    

class TwoWayList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.count = 0

    def is_empty(self):
        return self.head is None

    def append(self, data):
        new_Object = Object(data)
        new_Object.index = self.count
        self.count += 1
        if self.is_empty():
            self.head = new_Object
            self.tail = new_Object
        else:
            new_Object.prev = self.tail
            self.tail.next = new_Object
            self.tail = new_Object

    def prepend(self, data):
        new_Object = Object(data)
        new_Object.index = 0
        self.count += 1
        if self.is_empty():
            self.head = new_Object
            self.tail = new_Object
        else:
            new_Object.next = self.head
            self.head.prev = new_Object
            self.head = new_Object
            current = self.head.next
            while current:
                current.index += 1
                current = current.next

    def insert(self, data, index):
        new_Object = Object(data)
        new_Object.index = index
        self.count += 1
        current = self.head
        while current:
            if current.index == index:
                current.prev.next = new_Object
                new_Object.next = current
                new_Object.prev = current.prev
                current.prev = new_Object
                while current:
                    current.index += 1
                    current = current.next
                break
            current = current.next


    def show(self):
        current = self.head
        while current is not None:
            if current.index != 0:
                print("\t↑\n\t↓")
            current.show()
            current = current.next
        print("End of the list\n")
    
    def show_by_index(self, index):
        half = self.count / 2
        if index <= half:
            current = self.head
            while current.index != index:
                current = current.next
            current.show()
        else:
            current = self.tail
            while current.index != index:
                current = current.prev
            current.show()
    
    def find_element(self, index):
        half = self.count / 2
        if index <= half:
            current = self.head
            while current.index != index:
                current = current.next
        else:
            current = self.tail
            while current.index != index:
                current = current.prev
        return current
            
    def get_list_of_shortest_cadence(self):
        current = self.head
        min = current.data.years
        arr = np.array([])
        for i in range(self.count):
            if current.data.years < min:
                min = current.data.years
                arr = np.array([i])
            elif current.data.years == min:
                arr = np.append(arr, i)
            current = current.next
        list = TwoWayList()
        for i in range(len(arr)):
            list.append(self.find_element(arr[i]).data.get_data())
        return list
    
    def get_list_of_all_prsd_from_party(self, party):
        current = self.head
        list = TwoWayList()
        while current:
            if current.data.party == party:
                list.append(current.data.get_data())
            current = current.next
        return list
    
    def get_prsd_by_year(self, year):
        current = self.head
        list = TwoWayList()
        while current:
            if current.data.start <= year and year <= current.data.end:
                list.append(current.data.get_data())
            current = current.next
        if list.count == 1:
            return list.head
        if list.count > 1:
            return list
        print('wrong year')
        return None
    
    def delete_by_index(self, index):
        to_delete = self.find_element(index)
        if 0 < index and index < self.tail.index:
            to_delete.prev.next = to_delete.next
            to_delete.next.prev = to_delete.prev 
        elif index == 0:
            self.head = self.head.next
            self.head.prev = None
        elif index == self.tail.index:
            self.tail = self.tail.prev
            self.tail.next = None
        else:
            print('wrong index')
            return None
        current = to_delete.next
        while current:
            current.index -= 1
            current = current.next
        self.count -= 1
        return None


presidents = pd.read_csv('presidents.csv')
presidents['End'] = pd.to_numeric(presidents['RuleYears'].str[-4:]) 
presidents['Start'] = pd.to_numeric(presidents['RuleYears'].str[:4])
presidents['RuleYears'] = presidents['End'] - presidents['Start']
presidents.to_csv('prsd.csv')

for i in range(5):
    for j in range(len(presidents)):
        print(presidents.iloc[j,i],', ', end='')
    print('\n')

USA_prsd = TwoWayList()
for i in range(len(presidents)):
    USA_prsd.append(presidents.iloc[i,:5])

USA_prsd.insert(presidents.iloc[0,:5], 10)
#USA_prsd.show()
USA_prsd.delete_by_index(10)

#zad 1
#USA_prsd.get_list_of_shortest_cadence().show()

#zad 2
#USA_prsd.get_list_of_all_prsd_from_party('Republican').show()

#zad 3
#USA_prsd.get_prsd_by_year(1841).show()

#USA_prsd.show_by_index(10)
#USA_prsd.show_by_index(20)

#USA_prsd.show()
USA_prsd.delete_by_index(0)
#USA_prsd.show()
USA_prsd.delete_by_index(8)
#USA_prsd.show()

File: Data algorithms and structures/02 list/presidents.cpp
#include <iostream>
#include <string>
using namespace std;

class President{
public:
    string name;
    string party;
    int years;
    int start;
    int end;

    President(string n, string p, int y, int s, int e) : name(n), party(p), years(y), start(s), end(e) {}
    President() : years(0), name(""), party(""), start(0), end(0) {}

    void show(){
        cout<<"name: "<<name<<"\nparty:"<<party<<"\nyears:"<<years<<"\nstart:"<<start<<"\nend:"<<end<<endl;
    }
};

class Object{
public:
    President president;
    Object *next;
    Object *prev;
    int index;

    Object(President pr, Object *n, Object *p, int i) : president(pr), next(n), prev(p), index(i) {}
    Object(President pr, int i) : president(pr), index(i), next(nullptr), prev(nullptr) {}
    Object() : next(nullptr), prev(nullptr), president(President()), index(0) {}

    void show(){
        cout<<"index: "<<index<<endl;
        president.show();
    }
};

class TwoWayList{
public:
    Object* head;
    Object* tail;
    int count = 0;

    TwoWayList() : head(nullptr), tail(nullptr) {}

    void append(President prs){
        Object* newObj = new Object(prs, count);
        count++;
        if(head == nullptr){
            head = newObj;
            tail = newObj;
        }else{
            newObj->prev = tail;
            tail->next = newObj;
            tail = newObj;
        }
    }

    void show(){
        Object* current = head;
        while(current != nullptr){
            if(current->index != 0)
                cout<<"\t↑\n\t↓"<<endl;
            current->show();
            current = current->next;
        }
        cout<<"End of the list\n\n\n";
    }

    TwoWayList getShortestCadence() {
        Object *current = head;
        int min = current->president.years;
        for (int i = 0; i < count; i++) {
            if (current->president.years < min)
                min = current->president.years;
            current = current->next;
        }
        Object *current2 = head;
        TwoWayList list;
        for(int i = 0; i < count; i++) {
            if (current2->president.years == min)
                list.append(current2->president);
            current2 = current2->next;
        }
        return list;
    }

    TwoWayList getAllFromParty(string party){
        TwoWayList list;
        Object* current = head;
        for(int i = 0; i < count; i++){
            if(current->president.party == party)
                list.append(current->president);
            current = current->next;
        }
        return list;
    }

    TwoWayList getPrsdByYear(int year){
        TwoWayList list;
        Object* current = head;
        for(int i = 0; i < count; i++){
            if(current->president.start <= year && year <= current->president.end)
                list.append(current->president);
            current = current->next;
        }
        return list;
    }
};

int main() {
    string names[25] = {"George Washington","John Adams","Thomas Jefferson","James Madison","James Monroe","John Quincy Adams","Andrew Jackson","Martin Van Buren","William Henry Harrison","John Tyler","James K. Polk","Zachary Taylor","Millard Fillmore","Franklin Pierce","James Buchanan","Abraham Lincoln","Andrew Johnson","Ulysses S. Grant","Rutherford B. Hayes","James A. Garfield","Chester A. Arthur","Grover Cleveland","Benjamin Harrison","Grover Cleveland","William McKinley"};
    string party[25] = {"Unaffiliated","Federalist","Democratic-Republican","Democratic-Republican","Democratic-Republican","Democratic-Republican","Democratic","Democratic","Whig","Whig","Democratic","Whig","Whig","Democratic","Democratic","Republican","Democratic","Republican","Republican","Republican","Republican","Democratic","Republican","Democratic","Republican"};
    int start[25] = {1789 , 1797 , 1801 , 1809 , 1817 , 1825 , 1829 , 1837 , 1841 , 1841 , 1845 , 1849 , 1850 , 1853 , 1857 , 1861 , 1865 , 1869, 1877 , 1881 , 1881 , 1885 , 1889 , 1893 , 1897};
    int end[25] = {1797 , 1801 , 1809 , 1817 , 1825 , 1829 , 1837 , 1841 , 1841 , 1845 , 1849 , 1850 , 1853 , 1857 , 1861 , 1865 , 1869 , 1877, 1881 , 1881 , 1885 , 1889 , 1893 , 1897 , 1901};
    int ruleYears[25] = {8 , 4 , 8 , 8 , 8 , 4 , 8 , 4 , 0 , 4 , 4 , 1 , 3 , 4 , 4 , 4 , 4 , 8 , 4 , 0 , 4 , 4 , 4 , 4 , 4};

    TwoWayList usa;
    for(int i = 0; i < 25; i++)
        usa.append(President(names[i], party[i], ruleYears[i], start[i], end[i]));
    //usa.show();

    //zad 1
    TwoWayList shrCadence = usa.getShortestCadence();
    shrCadence.show();

    //zad 2
    TwoWayList republicans = usa.getAllFromParty("Republican");
    republicans.show();

    //zad 3
    TwoWayList prsd_1841 = usa.getPrsdByYear(1841);
    prsd_1841.show();

    return 0;
}
